daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesZoned.scala

```diff
@@ -17,13 +17,162 @@
+import org.apache.daffodil.util.Maybe._
+import org.apache.daffodil.util.MaybeDouble
 
-abstract class ZonedTextNumberPrim(e: ElementBase, guard: Boolean) extends UnimplementedPrimitive(e, guard) {
-  // lazy val parser: DaffodilParser = new ZonedTextNumberParser(e.elementRuntimeData)
+case class ConvertZonedCombinator(e: ElementBase, value: Gram, converter: Gram)
+  extends Terminal(e, !(value.isEmpty || converter.isEmpty)) {
+
+  lazy val parser = new ConvertZonedCombinatorParser(e.termRuntimeData, value.parser, converter.parser)
+
+  override lazy val unparser = new ConvertZonedCombinatorUnparser(e.termRuntimeData, value.unparser, converter.unparser)
+}
+
+abstract class ConvertZonedNumberPrim[S](e: ElementBase)
+  extends Terminal(e, true) {
+
+  def helper: ConvertZonedNumberParserUnparserHelperBase[S]
+
+  def numFormatFactory: ZonedFormatFactoryBase[S] = {
+    val h = helper
+
+    val pattern = {
+      val p = e.textNumberPattern
+
+      val noEscapedTicksRegex = """''""".r
+      val patternNoEscapedTicks = noEscapedTicksRegex.replaceAllIn(p, "")
+      val noQuotedRegex = """'[^']+'""".r
+      val patternNoQuoted = noQuotedRegex.replaceAllIn(patternNoEscapedTicks, "")
+
+      if (patternNoQuoted.contains("V")) {
+        e.notYetImplemented("textNumberPattern with V symbol")
+      }
+
+      if (patternNoQuoted.contains("P")) {
+        e.notYetImplemented("textNumberPattern with P symbol")
+      }
+
+      if (patternNoQuoted.contains("@")) {
+        e.SDE("The '@' symbol may not be used in textNumberPattern for textNumberRep='zoned'")
+      }
+
+      if (patternNoQuoted.contains(";")) {
+        e.SDE("Negative patterns may not be used in textNumberPattern for textNumberRep='zoned'")
+      }
+
+      e.primType match {
+        case PrimType.Double | PrimType.Float => e.SDE("textZonedFormat='zoned' does not support Doubles/Floats")
+        case PrimType.UnsignedLong | PrimType.UnsignedInt | PrimType.UnsignedShort | PrimType.UnsignedByte => {
+          if (e.textNumberCheckPolicy == TextNumberCheckPolicy.Lax) {
+            if ((patternNoQuoted.charAt(0) != '+') && (patternNoQuoted.charAt(patternNoQuoted.length - 1) != '+'))
+              e.SDE("textNumberPattern must have '+' at the beginning or the end of the pattern when textZonedFormat='zoned' and textNumberPolicy='lax' for unsigned numbers")
+          }
+        }
+        case _ => {
+          if ((patternNoQuoted.charAt(0) != '+') && (patternNoQuoted.charAt(patternNoQuoted.length - 1) != '+'))
+            e.SDE("textNumberPattern must have '+' at the beginning or the end of the pattern when textZonedFormat='zoned' for signed numbers")
+        }
+      }
+
+      // Load the pattern to make sure it is valid
+      try {
+        new DecimalFormat(p)
+      } catch {
+        case ex: IllegalArgumentException => e.SDE("Invalid textNumberPattern: " + ex.getMessage())
+      }
+
+      p
+    }
+
+    val (roundingIncrement: MaybeDouble, roundingMode) =
```

Comment:
This is how it was done for the standard text parser.  I'm OK with changing it as I'm not really sure what the purpose using MaybeDouble or One(e.textNumberRoundingMode) is. Was it there in case textNumberRoundingIncrement/Mode weren't defined by the schema?

Reply:
I was just saying the type MaybeDouble shouldn't be needed due to type inference. If this comes from  the standard text parser, it's probably not worth changing. In general, I might say maybe this should be a trait like RoundingTextNumber or something, and then the code could be shared between text standard and text zoned, but it's such little code I'm not sure addng a new trait it's worth it. All that to so, probably just ignore my above comment keep this as is.